core/pkg/server/sender.go

```diff
@@ -99,6 +100,12 @@ type Sender struct {
 	// runfilesUploader manages uploading a run's files
 	runfilesUploader runfiles.Uploader
 
+	// artifactsSaver manages artifact uploads
+	artifactsSaver *artifacts.ArtifactSaver
+
+	// artifactWG is a wait group for artifact-related goroutines
+	artifactWG sync.WaitGroup
```

Comment:
Yeah, I agree! The WaitGroup is necessary now because we process the result of saving in the Sender, which I wanted to avoid refactoring, but this work would probably be appropriate in the artifacts subpackage. Then we can remove the return value from `Save()` and add a `Finish()` method to the ArtifactSaveManager.

There are three bits of work that could be moved:

1. Error logging. This can happen in the ArtifactSaveManager, the same as for any other component of the internal process, like RunfilesUploader and FileStream.
2. Detecting the run code artifact. This might not be right for ArtifactSaveManager, but a higher level abstraction would do. I do think we need an "ArtifactManager" type that also coordinates linking and downloading; this can replace the "ArtifactSaveManager" which has a tiny implementation anyway.
3. Responding to the user process. We only ever do this in the Handler and Sender at the moment, so we'd have to establish a new pattern, but I think it makes sense for all artifact business logic to live in the artifacts package.